rust/main/chains/hyperlane-radix/src/provider/mod.rs

```diff
@@ -0,0 +1,200 @@
+use derive_new::new;
+use gateway_api_client::apis::configuration::Configuration as GatewayConfig;
+use gateway_api_client::models::{
+    CommittedTransactionInfo, GatewayStatusResponse, StateEntityDetailsRequest,
+    StateEntityDetailsResponse, StreamTransactionsRequest, StreamTransactionsResponse,
+    TransactionCommittedDetailsRequest, TransactionPreviewV2Request, TransactionPreviewV2Response,
+    TransactionStatusRequest, TransactionStatusResponse, TransactionSubmitRequest,
+    TransactionSubmitResponse,
+};
+use hyperlane_core::rpc_clients::BlockNumberGetter;
+use hyperlane_core::ChainResult;
+
+use crate::HyperlaneRadixError;
+
+mod fallback;
+mod radix;
+
+pub use {fallback::*, radix::*};
+
+/// Base Raidx provider
+/// defined the most basic methods the provider has to implement.
+/// Provides abstraction over the radix providers, that allows to implement features like Fallback behaviour
+#[async_trait]
+pub trait RadixGatewayProvider {
+    /// Gateway status
+    async fn gateway_status(&self) -> ChainResult<GatewayStatusResponse>;
+    /// Get committed transaction details
+    async fn transaction_committed(
+        &self,
+        tx_intent: TransactionCommittedDetailsRequest,
+    ) -> ChainResult<CommittedTransactionInfo>;
+    /// Submits a tx to the gateway
+    async fn submit_transaction(&self, tx: Vec<u8>) -> ChainResult<TransactionSubmitResponse>;
+    /// Get a preview of the tx
+    async fn transaction_preview(
+        &self,
+        request: TransactionPreviewV2Request,
+    ) -> ChainResult<TransactionPreviewV2Response>;
+    /// Get committed tx by various filters
+    async fn stream_txs(
+        &self,
+        request: StreamTransactionsRequest,
+    ) -> ChainResult<StreamTransactionsResponse>;
+    /// Get tx status
+    async fn transaction_status(
+        &self,
+        intent_hash: String,
+    ) -> ChainResult<TransactionStatusResponse>;
+    /// Get entity details
+    async fn entity_details(
+        &self,
+        request: StateEntityDetailsRequest,
+    ) -> ChainResult<StateEntityDetailsResponse>;
+}
+
+#[async_trait]
+/// Radix core provider
+pub trait RadixCoreProvider {
+    /// Call preview a contract method
+    async fn call_preview(
+        &self,
+        request: TransactionCallPreviewRequest,
+    ) -> ChainResult<TransactionCallPreviewResponse>;
+}
+
+/// Base Radix provider
+#[derive(new, Debug, Clone)]
+pub struct RadixBaseGatewayProvider {
+    gateway: GatewayConfig,
+}
+
+/// Base Radix provider
+#[derive(new, Debug, Clone)]
+pub struct RadixBaseCoreProvider {
+    core: CoreConfig,
+}
+
+#[async_trait]
+impl BlockNumberGetter for RadixBaseGatewayProvider {
+    /// Latest block number getter
+    async fn get_block_number(&self) -> ChainResult<u64> {
+        todo!()
+    }
+}
+
+#[async_trait]
+impl BlockNumberGetter for RadixBaseCoreProvider {
+    /// Latest block number getter
+    async fn get_block_number(&self) -> ChainResult<u64> {
+        todo!()
+    }
+}
```

Comment:
I think we need this methods implemented so that FallbackProvider can be functional.